### PR TITLE
Eng 9295 integration small with per block schema

### PR DESCRIPTION
**Pull request by akhanzode**
Per block schema changes.

### CODE REVIEW

File: src/ee/storage/ExportTupleStream.cpp [hunk 1]
```diff
@@ -95,13 +95,18 @@ size_t ExportTupleStream::appendTuple(int64_t lastCommittedSpHandle,
...
+    bool includeSchema = m_new;
     if (!m_currBlock) {
         extendBufferChain(m_defaultCapacity);
+        includeSchema = true;
     }
+
     // get schema related size
     size_t namesLength = computeSchemaSize(tableName, columnNames);
-    if (m_currBlock->remaining() < (tupleMaxLength + namesLength)) {
+    if ((m_currBlock->remaining() < (tupleMaxLength + namesLength)) ) {
```
Comment (luyangco) [inline src/ee/storage/ExportTupleStream.cpp]: Extra pair of parentheses.

File: src/frontend/org/voltdb/exportclient/ExportRow.java [hunk 2]
```diff
@@ -76,32 +78,103 @@ public String getPartitionColumnName() {
...
         boolean[] is_null = extractNullFlags(bb, columnCount);
 
-        String tableName = decodeString(bb);
         List<String> colNames = new ArrayList<>();
         List<Integer> colLengths = new ArrayList<>();
         List<VoltType> colTypes = new ArrayList<>();
+        String tableName = null;
+        if (hasSchema == 1) {
+            tableName = decodeString(bb);
+            colNames = new ArrayList<>();
```
Comment (luyangco) [inline src/frontend/org/voltdb/exportclient/ExportRow.java]: Double initialization.

File: src/frontend/org/voltdb/VoltTable.java [hunk 3]
```diff
@@ -508,6 +508,7 @@ public final void clearRowData() {
         m_buffer.position(m_rowStart);
         m_buffer.putInt(0);
         m_rowCount = 0;
+        m_activeRowIndex = INVALID_ROW_INDEX;
```
Comment (luyangco) [inline src/frontend/org/voltdb/VoltTable.java]: I don't understand this line. Is any reason to add that?

File: src/frontend/org/voltdb/exportclient/ExportDecoderBase.java [hunk 4]
```diff
@@ -85,7 +85,7 @@ public ExportRowData(Object[] vals, Object pval, int pid) {
      * @throws IOException
      */
     protected ExportRowData decodeRow(byte[] rowData) throws IOException {
-        ExportRow row = ExportRow.decodeRow(getPartition(), m_startTS, rowData);
+        ExportRow row = ExportRow.decodeRow(null, getPartition(), m_startTS, rowData);
```
Comment (luyangco) [inline src/frontend/org/voltdb/exportclient/ExportDecoderBase.java]: Why use `null` here? Does it guarantee to have schema in `rowData`?

File: src/frontend/org/voltdb/exportclient/ExportEncoder.java [hunk 5]
```diff
@@ -97,6 +98,76 @@
...
+                    hasSchema = false;
+                }
+
+                // write the non-null columns
+                for (int i = 0; i < colCount; i++) {
+                    if (!nullArray[i]) {
+                        encodeColumn(fs, i, table);
+                    }
+                }
+                System.out.println("Row done.");
```
Comment (luyangco) [inline src/frontend/org/voltdb/exportclient/ExportEncoder.java]: Get rid of debug statement.

File: src/frontend/org/voltdb/exportclient/ExportDecoderBase.java (hunk 4 above)
Comment (akhanzode) [reply src/frontend/org/voltdb/exportclient/ExportDecoderBase.java]: I will make legacy connector use schema from ADS.

File: src/frontend/org/voltdb/exportclient/ExportDecoderBase.java [hunk 6]
```diff
@@ -58,16 +70,27 @@ public RestartBlockException(String msg, boolean requestBackoff) {
     }
     protected final int m_partition;
     protected final long m_startTS;
+    //If true we have detected an old style connector.
     private boolean m_legacy = false;
+    //Only used for legacy connector which picks up schema from ADS
+    protected final ArrayList<VoltType> m_tableSchema = new ArrayList<>();
```
Comment (luyangco) [inline src/frontend/org/voltdb/exportclient/ExportDecoderBase.java]: For those member variables I suggest adding "legacy" prefix to their names.

File: src/frontend/org/voltdb/exportclient/ExportDecoderBase.java [hunk 7]
```diff
@@ -95,6 +118,95 @@ protected ExportRowData decodeRow(byte[] rowData) throws IOException {
         return new ExportRowData(row.values, row.partitionValue, row.partitionId);
     }
 
+    //This is for legacy connector.
+    public boolean writeRow(Object row[], CSVWriter writer, boolean skipinternal,
```
Comment (luyangco) [inline src/frontend/org/voltdb/exportclient/ExportDecoderBase.java]: For those legacy-only functions, could you add "legacy" prefix to differentiate them from the normal one?

File: src/ee/storage/ExportTupleStream.h [hunk 8]
```diff
@@ -83,24 +83,26 @@ class ExportTupleStream : public voltdb::TupleStreamBase {
             int partitionColumn,
             ExportTupleStream::Type type);
 
-    size_t computeOffsets(const TableTuple &tuple,size_t *rowHeaderSz) const;
-    size_t computeSchemaSize(const std::string &tableName, const std::vector<std::string> &columnNames) const;
+    size_t computeOffsets(const TableTuple &tuple, size_t *rowHeaderSz) const;
+    size_t computeSchemaSize(const std::string &tableName, const std::vector<std::string> &columnNames);
```
Comment (jhugg) [inline src/ee/storage/ExportTupleStream.h]: Not const anymore?

**Pull request merged**